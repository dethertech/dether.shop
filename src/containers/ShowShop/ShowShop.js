import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

/*
  Components
 */
import { ShopRecap, Button, TransactionFlow, Loader } from '../../components';
import { Padding } from '../../components/Spaces';

/*
  Translate module
 */
import tr from '../../translate';

/*
  Redux
 */
import {
  removeShop as removeShopAction,
  resetTransaction as resetTransactionAction,
  openNotificationModal as openNotificationModalAction,
  reloadShops as reloadShopsAction,
  hideShopWillAppearNotice as hideShopWillAppearNoticeAction,
} from '../../actions';

/*
  Helpers
 */
import { deleteShop as deleteShopHelper, getShop } from '../../helpers';
/*
  Constants
 */
import { notificationsTypes } from '../../constants';

/**
 * ShowShop container
 * @extends PureComponent
 */
export class ShowShop extends PureComponent {
  static propTypes = {
    shop: PropTypes.shape({
      name: PropTypes.string.isRequired,
      cat: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      opening: PropTypes.string.isRequired,
      lat: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
      lng: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    }).isRequired,
    isTransactionPending: PropTypes.bool.isRequired,
    deleteContractShop: PropTypes.func.isRequired,
    removeShopFromStore: PropTypes.func.isRequired,
    centerPosition: PropTypes.shape({}).isRequired,
    openNotificationModal: PropTypes.func.isRequired,
    reloadShops: PropTypes.func.isRequired,
    dispatchHideShopWillAppearNotice: PropTypes.func,
    shallIDisplayShopNotice: PropTypes.bool,
  };

  static defaultProps = {
    dispatchHideShopWillAppearNotice: () => null,
    shallIDisplayShopNotice: false,
  };

  state = {
    transactionSubmitted: false,
  };

  componentDidMount = () => {
    setTimeout(() => {
      this.props.dispatchHideShopWillAppearNotice();
    }, 60000);
  };

  checkTransaction = async () => {
    const {
      removeShopFromStore,
      resetTransaction,
      centerPosition,
      openNotificationModal,
      reloadShops,
    } = this.props;

    const shop = await getShop();
    if (!shop) {
      reloadShops(centerPosition);
      this.setState({ transactionSubmitted: false });
      resetTransaction();
      removeShopFromStore(shop);
      openNotificationModal({
        type: notificationsTypes.SUCCESS,
        message: tr('notifications.shop_deleted'),
      });
      return true;
    }
  };

  deleteShop = async () => {
    const { deleteContractShop } = this.props;

    const transaction = await deleteContractShop();
    return transaction.transactionHash;
  };

  handleError = () => {
    this.setState({ transactionSubmitted: false });
  };

  submitTransaction = () => this.setState({ transactionSubmitted: true });

  render = () => {
    const { shop, isTransactionPending, shallIDisplayShopNotice } = this.props;
    const { transactionSubmitted } = this.state;

    if (isTransactionPending || transactionSubmitted) {
      return (
        <TransactionFlow
          isTransaction
          loader={{
            title: tr('add_form_verification.loader_title'),
            message: tr('add_form_verification.loader_add_message'),
          }}
          sendTransaction={this.deleteShop}
          onError={this.handleError}
          shop={shop}
          checkTransaction={this.checkTransaction}
        />
      );
    }
    return (
      <Fragment>
        <ShopRecap {...shop} />
        <Button onClick={this.submitTransaction}>
          {tr('show_shop.delete_button')}
        </Button>
        {shallIDisplayShopNotice && (
          <Padding all="m">
            <div>
              <b>{tr('show_shop.shop_appear_shortly')}</b>
            </div>
            <div>
              <Loader />
            </div>
          </Padding>
        )}
      </Fragment>
    );
  };
}

const mapStateToProps = ({ shop, map, transaction }) => ({
  shop: shop.shop,
  isTransactionPending: !!transaction.pending,
  centerPosition: map.centerPosition,
  shallIDisplayShopNotice: shop.displayShopWillAppear,
});

const mapDispatchToProps = dispatch => ({
  deleteContractShop: deleteShopHelper,
  removeShopFromStore: bindActionCreators(removeShopAction, dispatch),
  resetTransaction: bindActionCreators(resetTransactionAction, dispatch),
  reloadShops: bindActionCreators(reloadShopsAction, dispatch),
  openNotificationModal: bindActionCreators(
    openNotificationModalAction,
    dispatch,
  ),
  dispatchHideShopWillAppearNotice: bindActionCreators(
    hideShopWillAppearNoticeAction,
    dispatch,
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(ShowShop);
