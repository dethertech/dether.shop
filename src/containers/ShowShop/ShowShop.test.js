import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ShowShop } from './ShowShop';

jest.mock('axios', () => ({
  CancelToken: {
    source: () => ({
      token: '',
    }),
  },
}));
describe('Containers::ShowShop', () => {

  const shop = {
    lat: '30',
    lng: '3',
    address: '40 rue Citadelle Paris',
    countryId: 'fr',
    postalCode: '75010',
    name: 'Magasin',
    description: 'Super Magasin',
    cat: 'cool',
    opening: '0000000'
  };

  const props = {
    shop,
    addDeleteShopTransaction: () => null,
    openNotificationModal: () => null,
    isTransactionPending: false,
    deleteContractShop: () => null,
    removeShopFromStore: () => null,
    transactionHash: "0xwefwerfer",
    endTransaction: () => null,
    fetchAll: () => null,
    centerPosition: { lat: 30, lng: 3 },
    reloadShops: () => null,
    dispatchHideShopWillAppearNotice: () => null,
    shallIDisplayShopNotice: false,
  };
  const component = props => <ShowShop {...props} />

  it('should match shallow snapshop', () => {
    expect(shallow(component(props))).toMatchSnapshot('ShowShop::shallow');
  });
  it('should match mounted snapshop', () => {
    expect(toJson(component(props))).toMatchSnapshot('ShowShop::mount');
  });
});
