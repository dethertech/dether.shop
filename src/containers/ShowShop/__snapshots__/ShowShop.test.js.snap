// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowShop::mount 1`] = `undefined`;

exports[`ShowShop::shallow 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowShop
    addDeleteShopTransaction={[Function]}
    centerPosition={
        Object {
            "lat": 30,
            "lng": 3,
          }
    }
    deleteContractShop={[Function]}
    dispatchHideShopWillAppearNotice={[Function]}
    endTransaction={[Function]}
    fetchAll={[Function]}
    isTransactionPending={false}
    openNotificationModal={[Function]}
    reloadShops={[Function]}
    removeShopFromStore={[Function]}
    shallIDisplayShopNotice={false}
    shop={
        Object {
            "address": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          }
    }
    transactionHash="0xwefwerfer"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ShopRecap
          address="40 rue Citadelle Paris"
          cat="cool"
          countryId="fr"
          description="Super Magasin"
          lat="30"
          licencePrice={null}
          lng="3"
          name="Magasin"
          opening="0000000"
          postalCode="75010"
/>,
        <Button
          disabled={false}
          fullWidth={false}
          isSmall={false}
          onClick={[Function]}
          theme="light"
          width={null}
>
          Delete Shop
</Button>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "address": "40 rue Citadelle Paris",
          "cat": "cool",
          "countryId": "fr",
          "description": "Super Magasin",
          "lat": "30",
          "licencePrice": null,
          "lng": "3",
          "name": "Magasin",
          "opening": "0000000",
          "postalCode": "75010",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Delete Shop",
          "disabled": false,
          "fullWidth": false,
          "isSmall": false,
          "onClick": [Function],
          "theme": "light",
          "width": null,
        },
        "ref": null,
        "rendered": "Delete Shop",
        "type": [Function],
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ShopRecap
            address="40 rue Citadelle Paris"
            cat="cool"
            countryId="fr"
            description="Super Magasin"
            lat="30"
            licencePrice={null}
            lng="3"
            name="Magasin"
            opening="0000000"
            postalCode="75010"
/>,
          <Button
            disabled={false}
            fullWidth={false}
            isSmall={false}
            onClick={[Function]}
            theme="light"
            width={null}
>
            Delete Shop
</Button>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "address": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "licencePrice": null,
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Delete Shop",
            "disabled": false,
            "fullWidth": false,
            "isSmall": false,
            "onClick": [Function],
            "theme": "light",
            "width": null,
          },
          "ref": null,
          "rendered": "Delete Shop",
          "type": [Function],
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
